### PR TITLE
add xpu for amp

### PR DESCRIPTION
**Pull request by jingxu10**
As support for Intel GPU has been upstreamed, this PR is to add the XPU-related contents to AMP doc.

cc @mcarilli @ptrblck @leslie-fang-intel @jgong5

### CODE REVIEW

File: docs/source/amp.rst
Comment (dvrogozh) [inline docs/source/amp.rst:277]: causes wrong rendering (header not recognized, see github rendered page view). Or you meant `float16` instead of `bfloat16`?
```suggestion
"""""""""""""""""""""""""""""""""""""""""
```

File: docs/source/amp.rst
Comment (dvrogozh) [inline docs/source/amp.rst:32]:
```suggestion
:class:`torch.autocast` is new in version `1.10`. :class:`torch.cpu.amp.autocast` and :class:`torch.cuda.amp.autocast` will be deprecated. 
```

however, you might wish to write like this to preserve current change history:
```
:class:`torch.autocast` and :class:`torch.cpu.amp.autocast` are new in version `1.10`.
:class:`torch.cpu.amp.autocast` and :class:`torch.cuda.amp.autocast` will be deprecated.
```

File: docs/source/amp.rst
Comment (dvrogozh) [inline docs/source/amp.rst:310]: Not currently supported by XPU [1]. I did not check whether other items in this list are supported by XPU backend - see comment below.

[1] https://github.com/intel/torch-xpu-ops/blob/65c61554aa1023ee99dce101dc743317d705b863/src/aten/XPUFallback.template#L162

File: docs/source/amp.rst
Comment (dvrogozh) [inline docs/source/amp.rst:300]: Not currently supported by XPU [1]. I did not check whether other items in this list are supported by XPU backend - see comment below.

[1] https://github.com/intel/torch-xpu-ops/blob/65c61554aa1023ee99dce101dc743317d705b863/src/aten/XPUFallback.template#L351

File: docs/source/amp.rst
Comment (dvrogozh) [inline docs/source/amp.rst:365]: This comment also covers tables in prev. sections... Some, actually most, of the operations listed here are currently not supported by XPU backend. As such this document will be confusing for the end user. I suggest that you need to document only those operations which are really supported by XPU. Having placeholders in documentation does not seem as a right thing to do.

| op | Supported? | Link | 
| --- | --- | --- |
| ``addcdiv`` | no | https://github.com/intel/torch-xpu-ops/blob/65c61554aa1023ee99dce101dc743317d705b863/src/aten/XPUFallback.template#L170 |
| ``addcmul`` | no | https://github.com/intel/torch-xpu-ops/blob/65c61554aa1023ee99dce101dc743317d705b863/src/aten/XPUFallback.template#L171 |
| ``atan2`` | no | https://github.com/intel/torch-xpu-ops/blob/65c61554aa1023ee99dce101dc743317d705b863/src/aten/XPUFallback.template#L181 |
| ``bilinear`` | unknown | |
| ``cross`` | unknown | |
| ``dot`` | no | https://github.com/intel/torch-xpu-ops/blob/65c61554aa1023ee99dce101dc743317d705b863/src/aten/XPUFallback.template#L213 |
| ``grid_sample`` | no | https://github.com/pytorch/pytorch/issues/127002 |
| ``index_put`` | yes | https://github.com/intel/torch-xpu-ops/blob/65c61554aa1023ee99dce101dc743317d705b863/yaml/xpu_functions.yaml#L245 |
| ``scatter_add`` | yes | https://github.com/intel/torch-xpu-ops/blob/65c61554aa1023ee99dce101dc743317d705b863/yaml/xpu_functions.yaml#L267 |
| ``tensordot`` | unknown | |

File: docs/source/amp.rst
Comment (dvrogozh) [inline docs/source/amp.rst:381]: `binary_cross_entropy` is not supported by XPU [1]. Not sure about `binary_cross_entropy_with_logits`.


[1] https://github.com/intel/torch-xpu-ops/blob/65c61554aa1023ee99dce101dc743317d705b863/src/aten/XPUFallback.template#L188

File: docs/source/amp.rst
Comment (dvrogozh) [inline docs/source/amp.rst:389]: sigmoid is currently not supported by XPU [1]. Not sure about other operations mentioned here.

[1] https://github.com/intel/torch-xpu-ops/blob/65c61554aa1023ee99dce101dc743317d705b863/src/aten/XPUFallback.template#L380

**Review by dvrogozh — CHANGES_REQUESTED** [review]
Overall, 2 things:
1. I believe that documentation should represent current state rather than future state. Thus, need to remove all mentioned operations and associated descriptions if they are not supported by XPU backend.
2. These lists of unsupported operations might be good source of ops to consider for prioritization. I suggest to think on filing issues at pytorch and torch-xpu-ops level highlighting missing ops affecting pytorch documentation.

File: docs/source/amp.rst
Comment (guangyey) [inline docs/source/amp.rst:32]: Let us keep this as it is because the deprecation warning has existed on line 29-30.

File: docs/source/amp.rst
Comment (EikanWang) [reply docs/source/amp.rst:381]: The sigmoid has been supported by XPU. Although its implementation is based on CPU and we will continuously optimize its performance by replacing its implementation with XPU kernel.

File: docs/source/amp.rst
Comment (EikanWang) [reply docs/source/amp.rst:389]: Ditto

**Comment by EikanWang** [comment]
@dvrogozh , Regarding the operations that we have explicitly utilized CPU implementation, it can run normally for Intel GPU with full functionality. Performance-wise, we will optimize these operations by replacing their implementations with XPU kernels. Therefore, I think we do not need to highlight these operations in the doc if it is only for functionality.

**Comment by dvrogozh** [comment]
> @dvrogozh , Regarding the operations that we have explicitly utilized CPU implementation, it can run normally for Intel GPU with full functionality.

@EikanWang : I am confused. These "explicit" operations are running on CPU or they are running on GPU?

**Comment by EikanWang** [comment]
> > @dvrogozh , Regarding the operations that we have explicitly utilized CPU implementation, it can run normally for Intel GPU with full functionality.
> 
> @EikanWang : I am confused. These "explicit" operations are running on CPU or they are running on GPU?

It runs on CPU. But I mean it is an internal detailed implementation from the operation support perspective. Compared to the other operations that are neither unimplemented by XPU nor in the template file, these operations can run normally without any error.

**Comment by dvrogozh** [comment]
> It runs on CPU. But I mean it is an internal detailed implementation from the operation support perspective.

End user expectation when using XPU backend is that operations will be executed on Intel GPU. It's good that some missing operations were plugged with CPU fallbacks to allow users run something from functional standpoint, but I do not think you can pretend that these operations are implemented for XPU. And to me it's definitely will be wrong to document these operations as if they are really supported by XPU (which as I said implies that they will run on Intel GPUs).

**Comment by EikanWang** [comment]
We are not saying whether these operations run on Intel GPU or not. We are talking about whether these operations can run normally under Intel GPU context. Regarding the detailed implementation, it is a choice per the backend. Take the MPS as an example, it also explicitly fallbacks some operations to CPU - https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/mps/MPSFallback.mm#L88-L96.

We implement more operations in SYCL for performance but not from the functionality perspective. We need to obtain promising performance at a particular moment. For now, the functionality is the priority.

**Comment by EikanWang** [comment]
@dvrogozh , I'm okay to remove these operations that explicitly fallbakcs to CPU from the AMP list first and add them back for 2.5 as we will implement these operations by then.

**Comment by guangyey** [comment]
@dvrogozh address your comments. I have already removed these CPU fallback ops from the list.

File: docs/source/amp.rst
Comment (dvrogozh) [inline docs/source/amp.rst:279]: Some of the items in this list are not in the template and supported ops lists. Does that mean they are not supported? `__matmul__` is the example.

File: docs/source/amp.rst
Comment (dvrogozh) [inline docs/source/amp.rst:279]: I do not think that some of these operations are currently actually supported by XPU. I don't see neither in explicit fallback list nor in a list of implemented functions. And associated tests are skipped. **Thus, I think this PR might be documenting something which is currently not supported which is bad and will confuse end users.** @jingxu10 can you, please, discuss supported ops with the dev team?

For example:
```
$ fgrep -rsn addbmm
test/xpu/run_test_with_skip.py:426:    "test_dtypes_addbmm_xpu",
test/xpu/run_test_with_skip.py:431:    "test_dtypes_baddbmm_xpu",
test/xpu/run_test_with_skip.py:483:    "test_noncontiguous_samples_addbmm_xpu_complex64",
test/xpu/run_test_with_skip.py:484:    "test_noncontiguous_samples_addbmm_xpu_float32",
test/xpu/run_test_with_skip.py:485:    "test_noncontiguous_samples_addbmm_xpu_int64",
test/xpu/run_test_with_skip.py:495:    "test_noncontiguous_samples_baddbmm_xpu_complex64",
test/xpu/run_test_with_skip.py:496:    "test_noncontiguous_samples_baddbmm_xpu_int64",
test/xpu/run_test_with_skip.py:567:    "test_numpy_ref_addbmm_xpu_complex128",
test/xpu/run_test_with_skip.py:568:    "test_numpy_ref_addbmm_xpu_float64",
test/xpu/run_test_with_skip.py:569:    "test_numpy_ref_addbmm_xpu_int64",
test/xpu/run_test_with_skip.py:571:    "test_out_addbmm_xpu_float32",
test/xpu/run_test_with_skip.py:574:    "test_out_baddbmm_xpu_float32",
test/xpu/run_test_with_skip.py:577:    "test_out_requires_grad_error_addbmm_xpu_complex64",
test/xpu/run_test_with_skip.py:581:    "test_out_requires_grad_error_baddbmm_xpu_complex64",
test/xpu/run_test_with_skip.py:617:    "test_out_warning_baddbmm_xpu",
test/xpu/run_test_with_skip.py:645:    "test_variant_consistency_eager_baddbmm_xpu_complex64",
test/xpu/run_test_with_skip.py:646:    "test_variant_consistency_eager_baddbmm_xpu_float32",
test/xpu/run_test_with_skip.py:710:    "test_conj_view_addbmm_xpu_complex64",
test/xpu/run_test_with_skip.py:711:    "test_neg_conj_view_addbmm_xpu_complex128",
test/xpu/run_test_with_skip.py:732:    "test_conj_view_baddbmm_xpu_complex64",
test/xpu/run_test_with_skip.py:798:    "test_neg_conj_view_baddbmm_xpu_complex128",
test/xpu/run_test_with_skip.py:852:    "test_neg_view_addbmm_xpu_float64",
test/xpu/run_test_with_skip.py:857:    "test_neg_view_baddbmm_xpu_float64",
```

cc: @EikanWang , @jgong5, @fengyuan14

**Review by dvrogozh — CHANGES_REQUESTED** [review]

File: docs/source/amp.rst
Comment (EikanWang) [reply docs/source/amp.rst:279]: @dvrogozh , sorry for the late reply. Regarding mm-related operators for XPU, we have upstreamed to the stock PyTorch

File: docs/source/amp.rst
Comment (EikanWang) [reply docs/source/amp.rst]: We have supported mm-related operations.

File: docs/source/amp.rst
Comment (jingxu10) [reply docs/source/amp.rst:279]: @dvrogozh any further comments? Can we merge the PR?

Comment (dvrogozh) [reply docs/source/amp.rst:279]: @EikanWang : mm operations are a subset of operations in this list. I am trying to raise a broader question: are all operations listed here implemented to run on XPU device as of 97d692eb8c4b3afab17700a2fd918adcea0cba45 which is current commit in https://github.com/pytorch/pytorch/blob/main/third_party/xpu.txt? I do not see some of them in the https://github.com/intel/torch-xpu-ops/blob/main/yaml/xpu_functions.yaml, but I might not be fully aware of the  design and, hence miss them. If you believe are all are implemented, I am fine with this PR. Otherwise I believe this list should be sanitized to only those ops which are actually supported to run on XPU.

Comment (EikanWang) [reply docs/source/amp.rst:279]: for matmul/conv, https://github.com/pytorch/pytorch/tree/6206da55ef1e2477eea9b0467c1f68d21cad65c0/aten/src/ATen/native/mkldnn/xpu FYI

**Comment by EikanWang** [comment]
The XPU operation list looks good to me. @jingxu10

File: docs/source/amp.rst
Comment (dvrogozh) [reply docs/source/amp.rst:279]: > for matmul/conv

Thank you.

**Review by dvrogozh — APPROVED** [review]
> The XPU operation list looks good to me. @jingxu10

Sounds good then. Approved.

**Comment by jingxu10** [comment]
Hi @svekars , this PR is ready for merge. Could you help to merge it? Thanks.

**Comment by jingxu10** [comment]
Hi @lezcano @Skylion007 , could you help to take a review on this PR? Thank you.

**Comment by lezcano** [comment]
I don't have expertise in this area, nor I know what's the current roadmap for this feature. Neither that @Skylion007. You'll need to find a core maintainer that is happy to shepherd this effort. cc @albanD for ideas.

**Comment by jingxu10** [comment]
Thanks @lezcano 
Hi @albanD , Could you help to get this PR a reivew? We would like to meet the docathon deadline. Thanks.

**Review by albanD — APPROVED** [review]
SGTM

**Comment by svekars** [comment]
@malfet @ezyang

**Comment by malfet** [comment]
Two remarks:
 - Not sure what are the goals of the Decathlon, but this is PR is about documenting new/experimental feature, which feels a bit of of scope for the docathon

I have two philosophical concerns with this PR:
 - Are there any XPU AMP test that validate something like that
 - I though the policy were not to document experimental features and/or mark them explicitly as experimental

**Comment by jingxu10** [comment]
@EikanWang  @chuanqi129  any insights to the comment above?

**Comment by EikanWang** [comment]
> * Are there any XPU AMP test that validate something like that

We added AMP unit test cases to PyTorch and AMP Dynamo-HuggingFace E2E tests to [torch-xpu-ops](https://github.com/intel/torch-xpu-ops) repo

AMP unit test cases:

- https://github.com/pytorch/pytorch/blob/9e8443b56f5a83877803be3ba43f7941841904c9/test/test_xpu.py#L314

AMP Dynamo-HuggingFace E2E tests:

- AMP-BF16: https://github.com/intel/torch-xpu-ops/actions/runs/9314873868
- AMP-FP16: https://github.com/intel/torch-xpu-ops/actions/runs/9314866046

> * I though the policy were not to document experimental features and/or mark them explicitly as experimental

Thanks for your suggestions, @malfet . @jingxu10 , let's mark them explicitly as experimental.

**Comment by jingxu10** [comment]
Hi @EikanWang  @malfet  Done in https://github.com/pytorch/pytorch/pull/127276/files#diff-e864a4a04d0b93580be487bdfd9de2262d7424286c77f936deb6bc7699c1a959R261-R262

**Comment by jingxu10** [comment]
Hi @malfet , any further concerns you have on this PR? Anything blocking merge? Thanks.

**Comment by jingxu10** [comment]
Hi @malfet , I need to cherry-pick this PR into 2.4 before July 1. Is there anything blocking merge of this PR? Thank you.

**Review by malfet — APPROVED** [review]

**Comment by jingxu10** [comment]
Hi @svekars , Could you help on the merge issue above? Thanks!

**Comment by malfet** [comment]
> I need to cherry-pick this PR into 2.4 before July 1. Is there anything blocking merge of this PR? Thank you.

Can you please elaborate why? This is a docs only change and, if I remember correctly we do not document experimental features in the releases.

**Comment by jingxu10** [comment]
@malfet  Thanks a lot!

**Comment by jingxu10** [comment]
> > I need to cherry-pick this PR into 2.4 before July 1. Is there anything blocking merge of this PR? Thank you.
> 
> Can you please elaborate why? This is a docs only change and, if I remember correctly we do not document experimental features in the releases.

@EikanWang any insights?

**Comment by EikanWang** [comment]
> Can you please elaborate why? This is a docs only change and, if I remember correctly we do not document experimental features in the releases.

The document only intends to help users better understand the features that XPU has supported and what features they can try on Intel GPUs for PT2.4. If we just get the document landed in the main branch, users may be confused with the XPU feature scope for PT2.4.

**Comment by jingxu10** [comment]
Hi @svekars @malfet the ci in progress seems to be hanging. Could you help to take a look?

**Pull request closed without merging**
